common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java

```diff
@@ -243,6 +247,104 @@ public static void load(String name, ClassLoader loader) {
         }
     }
 
+    /**
+     * Loading the native library through the specified ClassLoader.
+     * @param loader - The specified ClassLoader where the native library will be loaded into
+     * @param name - The native library path or name
+     * @param absolute - Whether the native library will be loaded by path or not
+     */
+    private static void loadLibrary(final ClassLoader loader, final String name, final boolean absolute) {
+        try {
+            // Make sure the helper is belong to the target ClassLoader.
+            final Class<?> newHelper = tryToLoadClass(loader, NativeLibraryUtil.class);
+            AccessController.doPrivileged(new PrivilegedAction<Object>() {
+                @Override
+                public Object run() {
+                    try {
+                        /*
+                         * Invoke the helper to load the native library, if succeed, then the native
+                         * library belong to the specified ClassLoader.
+                         */
+                        Method method = newHelper.getMethod("loadLibrary",
+                                new Class<?>[] { String.class, boolean.class });
+                        return method.invoke(null, name, absolute);
+                    } catch (Exception e) {
+                        throw new IllegalArgumentException("Load library failed!", e);
+                    }
+                }
+            });
+        } catch (Exception e) { // Should by pass the UnsatisfiedLinkError here!
+            logger.debug("Unable to load the library: " + name + '.', e);
+            NativeLibraryUtil.loadLibrary(name, absolute);
+        }
+    }
+
+    /**
+     * Try to load the helper Class into specified ClassLoader.
+     * @param loader - The specified ClassLoader
+     * @param helper - The helper Class
+     * @return A new helper Class defined in the specified ClassLoader.
+     * @throws ClassNotFoundException Helper class not found or loading failed
+     */
+    private static Class<?> tryToLoadClass(final ClassLoader loader, final Class<?> helper)
+            throws ClassNotFoundException {
+        try {
+            return loader.loadClass(helper.getName());
+        } catch (ClassNotFoundException e) {
+            // The helper class NOT found in target ClassLoader, we have to define the helper class.
+            final byte[] classBinary = classToByteArray(helper);
+            return AccessController.doPrivileged(new PrivilegedAction<Class<?>>() {
+                @Override
+                public Class<?> run() {
+                    try {
+                        // Define the helper class in the target ClassLoader,
+                        //  then we can call the helper to load the native library.
+                        Method defineClass = ClassLoader.class.getDeclaredMethod("defineClass", String.class,
```

Comment:
@jasontedor so I understand you right that you think its fine to pull in in terms of `SecurityManager` stuff ?